indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,566 @@
+        shutdownFutures.add(scheduleTaskShutdown(taskWorkItem));
+      }
+    }
+    controlThreadExecutor.shutdown();
+    try {
+      ListenableFuture<List<Void>> shutdownFuture = Futures.successfulAsList(shutdownFutures);
+      shutdownFuture.get();
+    }
+    catch (Exception e) {
+      LOGGER.error(e, "Encountered exception when stopping all tasks.");
+    }
+
+    final DateTime start = DateTimes.nowUtc();
+    final long gracefulShutdownMillis = taskConfig.getGracefulShutdownTimeout().toStandardDuration().getMillis();
+
+    LOGGER.info("Waiting up to %,dms for shutdown.", gracefulShutdownMillis);
+    if (gracefulShutdownMillis > 0) {
+      try {
+        final boolean terminated = controlThreadExecutor.awaitTermination(
+            gracefulShutdownMillis,
+            TimeUnit.MILLISECONDS
+        );
+        final long elapsed = System.currentTimeMillis() - start.getMillis();
+        if (terminated) {
+          LOGGER.info("Finished stopping in %,dms.", elapsed);
+        } else {
+          final Set<String> stillRunning;
+          synchronized (tasks) {
+            stillRunning = ImmutableSet.copyOf(tasks.keySet());
+          }
+          LOGGER.makeAlert("Failed to stop task threads")
+                .addData("stillRunning", stillRunning)
+                .addData("elapsed", elapsed)
+                .emit();
+
+          LOGGER.warn(
+              "Executor failed to stop after %,dms, not waiting for it! Tasks still running: [%s]",
+              elapsed,
+              Joiner.on("; ").join(stillRunning)
+          );
+        }
+      }
+      catch (InterruptedException e) {
+        LOGGER.warn(e, "Interrupted while waiting for executor to finish.");
+        Thread.currentThread().interrupt();
+      }
+    } else {
+      LOGGER.warn("Ran out of time, not waiting for executor to finish!");
+    }
+  }
+
+  @Override
+  public Collection<TaskRunnerWorkItem> getRunningTasks()
+  {
+    synchronized (tasks) {
+      final List<TaskRunnerWorkItem> ret = new ArrayList<>();
+      for (final ThreadingTaskRunnerWorkItem taskWorkItem : tasks.values()) {
+        if (taskWorkItem.getThread() != null) {
+          ret.add(taskWorkItem);
+        }
+      }
+      return ret;
+    }
+  }
+
+  @Override
+  public Collection<TaskRunnerWorkItem> getPendingTasks()
+  {
+    synchronized (tasks) {
+      final List<TaskRunnerWorkItem> ret = new ArrayList<>();
+      for (final ThreadingTaskRunnerWorkItem taskWorkItem : tasks.values()) {
+        if (taskWorkItem.getThread() == null) {
+          ret.add(taskWorkItem);
+        }
+      }
+      return ret;
+    }
+  }
+
+  @Nullable
+  @Override
+  public RunnerTaskState getRunnerTaskState(String taskId)
+  {
+    final ThreadingTaskRunnerWorkItem workItem = tasks.get(taskId);
+    if (workItem == null) {
+      return null;
+    } else {
+      if (workItem.getThread() == null) {
+        return RunnerTaskState.PENDING;
```

Comment:
can't it mean that task is finished as we set thread to null when task finishes ?

Reply:
I forget if I had the thread set to null after task finishes in earlier revisions, but currently the thread doesn't get nulled out after task completion